### PR TITLE
feat: add delegate to monitor training

### PR DESCRIPTION
**Pull request by ocworld**
It is needed to monitor and hook in each training iteration.

It's because how well training my model and how much better metrics such as auc in each boost round.

contacts: keunhyun.oh@ahnlab.com

### CODE REVIEW

**Comment by ocworld** [comment]
@AhnLab-OSS

File: src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala [hunk 1]
```diff
@@ -247,9 +262,22 @@ private object TrainUtils extends Serializable {
...
         }
+
         lightgbmlib.delete_doubleArray(evalResults)
+
+        Option(Map(results:_*))
+      } else {
+        None
       }
+
+      if(delegate.isDefined) {
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala]: nit: space after if

File: src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala [hunk 2]
```diff
@@ -199,7 +199,14 @@ private object TrainUtils extends Serializable {
     val bestScore = new Array[Double](evalCounts)
     val bestScores = new Array[Array[Double]](evalCounts)
     val bestIter = new Array[Int](evalCounts)
+    val delegate = trainParams.delegate
+    val partitionId = TaskContext.getPartitionId
     while (!isFinished && iters < trainParams.numIterations) {
+
+      if(delegate.isDefined) {
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala]: nit: space after if

File: src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMParams.scala [hunk 3]
```diff
@@ -325,4 +325,11 @@ trait LightGBMParams extends Wrappable with DefaultParamsWritable with HasWeight
 
   def getMinDataInLeaf: Int = $(minDataInLeaf)
   def setMinDataInLeaf(value: Int): this.type = set(minDataInLeaf, value)
+
+  var delegate: Option[LightGBMDelegate] = None
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMParams.scala]: we should probably figure out how to make this inherit from param and expose in python layer, but maybe not for this PR

File: src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala (hunk 1 above)
Comment (ocworld) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala]: it is fixed

File: src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala (hunk 2 above)
Comment (ocworld) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala]: it is fixed :)

**Pull request merged**